### ISSUE TITLE
Callback order

### ISSUE THREAD

**Issue by Whitetigerswt** [Zeex/sampgdk#132]
Hey,

I'm using this filterscript: 
```cpp
#include <a_samp>

public OnPlayerConnect(playerid)
{
	printf("OnPlayerConnect(%d) - Filterscript PAWN", playerid);
    return 1;
}
```

This gamemode:

```cpp
#include <a_samp>

public OnPlayerConnect(playerid)
{
	printf("OnPlayerConnect - Gamemode PAWN");

	return 1;
}

main()
{
	
}
```

and this as a sampGDK plugin:
```cpp
PLUGIN_EXPORT bool PLUGIN_CALL OnPlayerConnect(int playerid)
{
    logprintf("OnPlayerConnect - Plugin");
    return 1;
}
```

When loading all, the callbacks call in this order:
```
[23:02:11] OnPlayerConnect(0) - Filterscript PAWN
[23:02:11] OnPlayerConnect - Plugin
[23:02:11] OnPlayerConnect - Gamemode PAWN
```

(On windows 0.3z-R4, if that matters...)

Why is the plugin not executing before ALL PAWN scripts? - Or is there some setting I can give to make the plugin have priority over all scripts? (or less priority?) 

This is causing a problem for me, in that I initialize something in OnPlayerConnect for each player in the plugin, but causes crashes because my filterscript is trying to do something with my plugin's functions before it's initialized.

It seems that https://github.com/Zeex/sampgdk/wiki/Callback-return-values indicates the plugin should have priority over ALL pawn scripts, but it executes callbacks AFTER filterscripts.

**Comment by ikkentim** [comment]
Not really 
`
(...) to stop the same callback from being called in other running plugins (if any) or the gamemode.
`
No mention of filterscripts in there.

**Comment by Whitetigerswt** [comment]
Another thing I noticed in trying to work around this issue:

OnPublicCall AMX *amx param is always the gamemodes AMX param, which makes you wonder, why is the amx param even passed since it's always the gamemode's amx?

**Comment by ikkentim** [comment]
amx is passed to use, for example, with amx_GetAddr which you need when looking up a string parameter.

**Comment by Zeex** [comment]
Yes, currently there's no way to catch publics in filterscripts, the `amx` parameter in OnPublicCall exists merely for convenience so that plugins don't need to detect/remember which AMX is the gamemode. Perhaps I'll extend OnPublicCall to support filterscripts, I need to think about it for a little (to not break something again).

If you need this urgenly I think you could just edit the amx_Exec() hook in [internal/amxhooks.c](https://github.com/Zeex/sampgdk/blob/master/lib/sampgdk/internal/amxhooks.c#L160-237), removing the `amx` checks [here](https://github.com/Zeex/sampgdk/blob/a2606da62ecb0c5351ac081dc777de8d397f973b/lib/sampgdk/internal/amxhooks.c#L193-194) and replacing `main_amx` in [this block](https://github.com/Zeex/sampgdk/blob/a2606da62ecb0c5351ac081dc777de8d397f973b/lib/sampgdk/internal/amxhooks.c#L200-203) with `amx`. The latter part would beak public calls in fakeamx, so if you're using `HTTP()` you'd also need to make sure you don't do index translation for it [here](https://github.com/Zeex/sampgdk/blob/a2606da62ecb0c5351ac081dc777de8d397f973b/lib/sampgdk/internal/amxhooks.c#L197).

So basically you'll be able to catch callbacks in both gamemode *and* filterscripts, except if they don't exist in filterscripts they won't be executed (unlike the gamemode), which is from what I understood would be sufficient for your purposes. This assumes that you're using the static version of the library because you don't want other plugins to do the same.
